### PR TITLE
add sparse creation support for pytorch

### PR DESCRIPTION
**Pull request by lanking520**
## Description ##
Add COO sparse NDArray creation for pytorch

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- [ ] Code is well-documented: 
    - For user-facing API changes, Java doc has been updated. 
    - For new examples, README.md is added to explain the what the example does.
- [ ] To the my best knowledge, [examples](https://github.com/awslabs/djl/tree/master/examples) and [jupyter notebooks](https://github.com/awslabs/djl/tree/master/jupyter) are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, Java doc)
- [ ] Feature2, tests, (and when applicable, Java doc)

### CODE REVIEW

File: pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDManager.java [hunk 1]
```diff
@@ -118,13 +118,24 @@ public PtNDArray create(Buffer data, Shape shape, DataType dataType) {
...
     public NDArray createRowSparse(Buffer data, Shape dataShape, long[] indices, Shape shape) {
-        throw new UnsupportedOperationException("Not implemented");
+        throw new UnsupportedOperationException("PyTorch doesn't support Row Sparse");
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public NDArray createCOO(Buffer data, long[] indices, Shape indicesShape, Shape shape) {
+        // length should be the same as indices dim 1
+        try (NDArray valueNd = create(data, new Shape(indicesShape.get(1)))) {
```
Comment (stu1130) [inline pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDManager.java:134]:
```suggestion
        try (NDArray valueNd = create(data, new Shape(indicesShape.get(indicesShape.dimension() - 1)))) {
```
1 only applied for 2-D

File: pytorch/pytorch-native/src/ai_djl_pytorch_jni_PyTorchLibrary_torch_creation.cc [hunk 2]
```diff
@@ -141,3 +141,18 @@ JNIEXPORT jobject JNICALL Java_ai_djl_pytorch_jni_PyTorchLibrary_torchEye(JNIEnv
   return utils::CreatePointer<torch::Tensor>(env, tensor_ptr);
   API_END_RETURN();
 }
+
+JNIEXPORT jobject JNICALL Java_ai_djl_pytorch_jni_PyTorchLibrary_torchSparseCOO(JNIEnv* env, jobject jthis, jlongArray jshape,
+    jobject jindices, jobject jvalue, jboolean jrequired_grad){
+    API_BEGIN()
+        const auto* indices_ptr = utils::GetPointerFromJHandle<const torch::Tensor>(env, jindices);
```
Comment (stu1130) [inline pytorch/pytorch-native/src/ai_djl_pytorch_jni_PyTorchLibrary_torch_creation.cc:148]: you can run `./gradlew fC` to format the code

File: pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDManager.java (hunk 1 above)
Comment (lanking520) [reply pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDManager.java:134]: In 3D, the indices shape will be in (3, num_of_elements)

File: api/src/main/java/ai/djl/ndarray/NDManager.java [hunk 3]
```diff
@@ -584,6 +584,34 @@ default NDArray createRowSparse(
...
+    /**
+     * Creates a 2D Coordinate Format (COO) Matrix.
+     *
+     * @param data the data to set for the Coordinate format {@link NDArray}
+     * @param row the row coordinate of the data
+     * @param column the column coordinate of the data
+     * @param shape the {@link Shape} of the {@link NDArray}
+     * @return a new instance of {@link NDArray}
+     */
+    default NDArray createCOO(Buffer data, long[] row, long[] column, Shape shape) {
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/ndarray/NDManager.java:596]: createCoo

File: api/src/main/java/ai/djl/ndarray/NDManager.java [hunk 4]
```diff
@@ -584,6 +584,34 @@ default NDArray createRowSparse(
...
+    /**
+     * Creates a 2D Coordinate Format (COO) Matrix.
+     *
+     * @param data the data to set for the Coordinate format {@link NDArray}
+     * @param row the row coordinate of the data
+     * @param column the column coordinate of the data
+     * @param shape the {@link Shape} of the {@link NDArray}
+     * @return a new instance of {@link NDArray}
+     */
+    default NDArray createCoo(Buffer data, long[] row, long[] column, Shape shape) {
```
Comment (stu1130) [inline api/src/main/java/ai/djl/ndarray/NDManager.java:596]: general though, do we want to use 2d array to represent indices?

File: integration/src/main/java/ai/djl/integration/tests/ndarray/NDArrayCreationOpTest.java [hunk 5]
```diff
@@ -94,12 +94,33 @@ public void testCreateRowSparseMatrix() {
         }
     }
 
+    @Test
+    public void testCreateCOOMatrix() {
```
Comment (stu1130) [inline integration/src/main/java/ai/djl/integration/tests/ndarray/NDArrayCreationOpTest.java:98]: nit: testCreateCooMatrix

File: pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/jni/PyTorchLibrary.java [hunk 6]
```diff
@@ -75,6 +75,9 @@ native Pointer torchZerosLike(
     native Pointer torchOnesLike(
             Pointer handle, int dType, int layout, int[] device, boolean requiredGrad);
 
+    native Pointer torchSparseCOO(
```
Comment (stu1130) [inline pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/jni/PyTorchLibrary.java:78]:
```suggestion
    native Pointer torchSparseCoo(
```

**Review by stu1130 — APPROVED** [review]

**Review by frankfliu — APPROVED** [review]

**Pull request merged**